### PR TITLE
CORE-8945 Add `POST /path-list-creator`

### PR DESCRIPTION
**Pull request by psarando**
This PR adds the `POST /path-list-creator` endpoint, which will generate an HT Path List file for all paths under the given set of folder paths, saving the results to the given destination file.

The paths for folders given in the request will not be included in the results, but if any file paths are included in the request, those paths will also be processed according to the other given parameters:
* `recursive`: When set to true, then paths under all subfolders will also be processed recursively.
* `folders-only`: When set to true, then only paths for folders will be included in the results, otherwise only file paths will be included.
* `name-pattern`: Specifying a regex pattern will filter the results to include only file/folder names that match this pattern.
* `info-type`: Specifying one or more info-types will filter the results to include only files with one of those info-types.

### CODE REVIEW

File: src/data_info/services/path_lists.clj [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+   Throws an error if the filtering params result in no matching paths."
+  [{:keys [user dest name-pattern info-type folders-only recursive]} {:keys [paths]}]
+  (irods/with-jargon-exceptions :client-user user [cm]
+    (validators/user-exists cm user)
+    (let [dest-dir (ft/dirname dest)]
+      (validators/path-exists cm dest-dir)
+      (validators/path-writeable cm user dest-dir)
+      (validators/path-not-exists cm dest)
+      (validators/all-paths-exist cm paths)
+      (validators/all-paths-readable cm user paths)
```
Comment (ianmcorvidae) [inline src/data_info/services/path_lists.clj]: It might be nice to be explicit, but since you're using a client user, you shouldn't need this line because the `exists` check one line above will already say it doesn't exist if it's not visible to the user.

File: src/data_info/services/path_lists.clj [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+(defn create-path-list
+  "Creates an HT Path List file from the given set of paths and filtering params.
+   The resulting HT Path List file will contain only file or only folder paths (depending on the `folders-only` param),
+   but no paths of folders in the initial given list will be included.
+   If `recursive` is true, then all subfolders (plus all their files and subfolders)
+   of any given folder paths are parsed and filtered as well.
+   Throws an error if the filtering params result in no matching paths."
+  [{:keys [user dest name-pattern info-type folders-only recursive]} {:keys [paths]}]
+  (irods/with-jargon-exceptions :client-user user [cm]
+    (validators/user-exists cm user)
```
Comment (ianmcorvidae) [inline src/data_info/services/path_lists.clj]: You shouldn't need this, since you're using a client user. If the client user doesn't exist, `with-jargon` will already have failed out with the same error.

File: src/data_info/services/path_lists.clj [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+         folders-only?
+         (not folders-only?))))
+
+(defn- paths->ht-path-list
+  "Filters the given paths and returns a string of these paths appended to an HT Path List header.
+   Throws an error if the filtering params result in no matching paths."
+  [cm user name-pattern info-types folders-only? recursive? paths]
+  (let [{folder-paths true file-paths false} (group-by (partial item-info/is-dir? cm) paths)
+        files          (->> file-paths
+                            (map (partial stat/path-stat cm user))
```
Comment (ianmcorvidae) [inline src/data_info/services/path_lists.clj]: I'm betting you can add some filters here to make things snappier. The `keep-top-level-file?` filter seems to care only about `infoType`, `keep-data-item?` seems to want `label` and `permission`, and `stat-is-dir?` wants `type`. Then, you filter down to just path. Since this is in data-info, you should be able to do `(map #(stat/path-stat cm user % :filter-include [:path :infoType :label :permission :type]))` or so, and thus save some bits of processing (particularly content-type).

File: src/data_info/services/path_lists.clj [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+       (label-matches? name-pattern label)
+       (if (stat-is-dir? data-item)
+         folders-only?
+         (not folders-only?))))
+
+(defn- paths->ht-path-list
+  "Filters the given paths and returns a string of these paths appended to an HT Path List header.
+   Throws an error if the filtering params result in no matching paths."
+  [cm user name-pattern info-types folders-only? recursive? paths]
+  (let [{folder-paths true file-paths false} (group-by (partial item-info/is-dir? cm) paths)
```
Comment (ianmcorvidae) [inline src/data_info/services/path_lists.clj]: This `item-info/is-dir?` will possibly slow things down for you. With this, you're basically already doing a decent chunk of a `stat` (the same call that `is-dir?` uses provides path, type, created/modified dates, file size, and checksum). I'm not sure if there's a sensible way to rewrite this to just fetch stats in the first place to reduce processing further down, though. The section below wants infoType, which would only apply to files so would be fine to add, but it also wants permission which is an extra iRODS call that only the files section below looks like it uses. Anyway, food for thought maybe?

File: src/data_info/services/path_lists.clj [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  (or (string/blank? name-pattern)
+      (re-find (re-pattern name-pattern) label)))
+
+(defn- keep-data-item?
+  "Returns true if the given `data-item` has permissions set,
+   its label passes the `label-matches?` check,
+   and if it's a file or folder depending on the given `folders-only?` flag."
+  [name-pattern folders-only? {:keys [label permission] :as data-item}]
+  (and permission
+       (label-matches? name-pattern label)
```
Comment (ianmcorvidae) [inline src/data_info/services/path_lists.clj]: I don't know if this is a real problem, but note that label isn't exactly the same as the basename of the path in three cases: the user's trash, `/iplant/home`, and `/iplant/home/shared`. Those might all constitute folders we should just outright disallow in this endpoint? not sure

File: src/data_info/services/path_lists.clj [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    (validators/user-exists cm user)
+    (let [dest-dir (ft/dirname dest)]
+      (validators/path-exists cm dest-dir)
+      (validators/path-writeable cm user dest-dir)
+      (validators/path-not-exists cm dest)
+      (validators/all-paths-exist cm paths)
+      (validators/all-paths-readable cm user paths)
+
+      (with-in-str (paths->ht-path-list cm user name-pattern info-type folders-only recursive paths)
+                   {:file (stat/decorate-stat cm user (copy-stream cm *in* user dest) (stat/process-filters nil nil))}))))
```
Comment (ianmcorvidae) [inline src/data_info/services/path_lists.clj]: Small thing, but would it be worth setting the `ht-analysis-path-list` info-type on this file right after uploading it? You could do that with `(clj-jargon.metadata/set-metadata cm path (cfg/type-detect-type-attribute) "ht-analysis-path-list" "")` (or if you don't mind munging things a bit, you could probably mess with `data-info.services.filetypes/add-type` to make it not always `with-jargon`/validate/etc). If you do, you'd want to do it before the `decorate-stat` call.

File: src/data_info/services/path_lists.clj (hunk 5 above)
Comment (psarando) [inline src/data_info/services/path_lists.clj]: That's probably a good idea. 👍

File: src/data_info/services/path_lists.clj (hunk 4 above)
Comment (psarando) [inline src/data_info/services/path_lists.clj]: Actually, in testing a request with paths of 40 folders and 40 files, calling `path-stat` with `filter-include` on those top-level folders as suggested (instead of using `item-info/is-dir?` on all paths then `path-stat` only on the top-level files) actually slowed down the endpoint by at least 1 second. Probably because of all the extra calls to include permissions that are not needed for the top-level folders, since the folder paths given in the request are not included in the results.

The biggest performance boost in this endpoint is using direct ICAT queries to get the listings (possibly recursive) for those requested folder paths (see the private `folder-listing` function above).

The typical use case for this endpoint is for requests with 1 folder path, or maybe a few folder paths (as if expanding paths under a folder into an HT Path List file). Since I implemented the endpoint to handle multiple paths, I thought it would be convenient for the user to also give them the option of including file paths in the request.

File: src/data_info/services/path_lists.clj (hunk 3 above)
Comment (psarando) [inline src/data_info/services/path_lists.clj]: Although I'm not expecting this endpoint to process too many file paths from the request, this is still a good idea in case it needs to 👍

File: src/data_info/services/path_lists.clj (hunk 4 above)
Comment (ianmcorvidae) [inline src/data_info/services/path_lists.clj]: Yeah -- my initial comment wasn't really very clear. What I was thinking of, and this is quite possibly premature optimization, was, instead of calling `item-info/is-dir?`, call `item-info/stat`, and then call out to `decorate-stat` in the file section below to get the extra info/filter out unneeded stuff. Then, you're only doing one `objStat` per top level thing passed in. As a bonus, if there's anything about that which can be used in the folder side of things, then you'd already have access to it (but I didn't anticipate that being a huge gain, by any means, since the icat queries are bulk operation sorts of things like all of this probably *should* be).

in short: when I said `stat` I should have been clear I meant you could do an `item-info/stat` here, rather than `data-info.services.stat/path-stat`.

In any case for the typical use case this is nearly irrelevant, and might be confusing (not sure), so it's your call if you want to do it!

**Comment by psarando** [comment]
Thanks for the review! 👍
